multicluster/service-mirror/cluster_watcher.go

```diff
@@ -1304,14 +1606,191 @@ func (rcsw *RemoteClusterServiceWatcher) isRemoteDiscovery(l map[string]string)
 	// Treat an empty remoteDiscoverySelector as "Nothing" instead of
 	// "Everything" so that when the remoteDiscoverySelector field is unset, we
 	// don't export all Services.
+	if rcsw.link.RemoteDiscoverySelector == nil {
+		return false
+	}
 	if len(rcsw.link.RemoteDiscoverySelector.MatchExpressions)+len(rcsw.link.RemoteDiscoverySelector.MatchLabels) == 0 {
 		return false
 	}
-	remoteDiscoverySelector, err := metav1.LabelSelectorAsSelector(&rcsw.link.RemoteDiscoverySelector)
+	remoteDiscoverySelector, err := metav1.LabelSelectorAsSelector(rcsw.link.RemoteDiscoverySelector)
 	if err != nil {
 		rcsw.log.Errorf("Invalid selector: %s", err)
 		return false
 	}
 
 	return remoteDiscoverySelector.Matches(labels.Set(l))
 }
+
+func (rcsw *RemoteClusterServiceWatcher) isFederatedServiceMember(l map[string]string) bool {
+	// Treat an empty federatedServiceSelector as "Nothing" instead of
+	// "Everything" so that when the federatedServiceSelector field is unset, we
+	// don't export all Services.
+	if rcsw.link.FederatedServiceSelector == nil {
+		return false
+	}
+	if len(rcsw.link.FederatedServiceSelector.MatchExpressions)+len(rcsw.link.FederatedServiceSelector.MatchLabels) == 0 {
+		return false
+	}
+	federatedServiceSelector, err := metav1.LabelSelectorAsSelector(rcsw.link.FederatedServiceSelector)
+	if err != nil {
+		rcsw.log.Errorf("Invalid selector: %s", err)
+		return false
+	}
+
+	return federatedServiceSelector.Matches(labels.Set(l))
+}
+
+func (rcsw *RemoteClusterServiceWatcher) updateLinkMirrorStatus(remoteName, namespace string, condition map[string]interface{}) {
+	err := rcsw.updateLinkStatus("mirrorServices", remoteName, namespace, condition)
+	if err != nil {
+		rcsw.log.Errorf("Failed to update link status: %s", err)
+	}
+}
+
+func (rcsw *RemoteClusterServiceWatcher) updateLinkFederatedStatus(remoteName, namespace string, condition map[string]interface{}) {
+	err := rcsw.updateLinkStatus("federatedServices", remoteName, namespace, condition)
+	if err != nil {
+		rcsw.log.Errorf("Failed to update link status: %s", err)
+	}
+}
+
+func (rcsw *RemoteClusterServiceWatcher) updateLinkStatus(statusSection, remoteName, namespace string, condition map[string]interface{}) error {
```

Comment:
Unfortunately, we use the dynamic client for interacting with Link resources instead of the generated binding.  I think we should migrate off the dynamic client and replace it with the generated bindings, but I didn't want to block this work on that.

Reply:
Yeah, that makes sense!